src/connector/src/source/kafka/enumerator/client.rs

```diff
@@ -94,36 +105,67 @@ impl SplitEnumerator for KafkaSplitEnumerator {
             scan_start_offset = KafkaEnumeratorOffset::Timestamp(time_offset)
         }
 
-        // don't need kafka metrics from enumerator
-        let ctx_common = KafkaContextCommon::new(
-            broker_rewrite_map,
-            None,
-            None,
-            properties.aws_auth_props,
-            common_props.is_aws_msk_iam(),
-        )
-        .await?;
-        let client_ctx = RwConsumerContext::new(ctx_common);
-        let client: BaseConsumer<RwConsumerContext> =
-            config.create_with_context(client_ctx).await?;
-
-        // Note that before any SASL/OAUTHBEARER broker connection can succeed the application must call
-        // rd_kafka_oauthbearer_set_token() once – either directly or, more typically, by invoking either
-        // rd_kafka_poll(), rd_kafka_consumer_poll(), rd_kafka_queue_poll(), etc, in order to cause retrieval
-        // of an initial token to occur.
-        // https://docs.confluent.io/platform/current/clients/librdkafka/html/rdkafka_8h.html#a988395722598f63396d7a1bedb22adaf
-        if common_props.is_aws_msk_iam() {
-            #[cfg(not(madsim))]
-            client.poll(Duration::from_secs(10)); // note: this is a blocking call
-            #[cfg(madsim)]
-            client.poll(Duration::from_secs(10)).await;
+        let kafka_client: Arc<BaseConsumer<RwConsumerContext>>;
+        let mut shared_client_guard = SHARED_KAFKA_CLIENT.lock().await;
+        if let Some(item) = shared_client_guard.get_mut(&connection_hash) {
+            tracing::info!(
+                "reusing kafka client for connection hash {}, to broker {}",
+                connection_hash,
+                broker_address
+            );
+            kafka_client = item.client.clone();
+            item.ref_count += 1;
+            drop(shared_client_guard);
+        } else {
+            // drop the guard and acquire a new one to avoid a 10s blocking call
+            drop(shared_client_guard);
```

Comment:
Will this happen?

1. Caller A tries to get a client for a connection, but cache missed, so it takes some time for calller A to build the client.
2. During it, caller B tries to get a client for the same connection, but cache missed, so it also builds a client.
3. Caller A built and inserted the client to the map and set ref_count = 1
4. One second later, caller B also inserted the client to the map and set ref_count = 1, **causing the client of caller A to leak and will never be dropped.**

Reply:
I would recommend to use `moka` for this case.

```rust
use moka::future::Cache;
```

Particularly, a caching structure should handle these concurrent gets correctly by letting the caller B blocks until caller A completes its operation and insert back the cached item i.e. the Kafka Client.